src/colmap/estimators/covariance.cc

```diff
@@ -0,0 +1,477 @@
+    Eigen::MatrixXd subMatrix = subMatrix_sparse;
+    subMatrix +=
+        lambda_ * Eigen::MatrixXd::Identity(subMatrix.rows(), subMatrix.cols());
+    Eigen::MatrixXd subMatrix_inv = subMatrix.inverse();
+    // update matrix
+    for (int i = 0; i < num_params_point; ++i) {
+      for (int j = 0; j < num_params_point; ++j) {
+        H_pp_inv.coeffRef(counter_p + i, counter_p + j) = subMatrix_inv(i, j);
+      }
+    }
+    counter_p += num_params_point;
+  }
+  H_pp_inv.makeCompressed();
+  S_matrix_ = H_cc - H_cp * H_pp_inv * H_pc;
+}
+
+bool BundleAdjustmentCovarianceEstimator::HasValidSchurComplement() const {
+  return S_matrix_.size() != 0;
+}
+
+bool BundleAdjustmentCovarianceEstimator::ComputeFull() {
+  if (!HasValidSchurComplement()) {
+    ComputeSchurComplement();
+  }
+  Eigen::MatrixXd S_dense = S_matrix_;
+  LOG(INFO) << StringPrintf(
+      "Inverting Schur complement for all variables except for 3D points (n = "
+      "%d)",
+      num_params_poses_ + num_params_other_variables_);
+  Eigen::FullPivLU<Eigen::MatrixXd> luOfS_dense(S_dense);
+  if (luOfS_dense.rank() < S_dense.rows()) {
+    LOG(INFO) << StringPrintf(
+        "Unable to compute covariance. The Schur complement on all variables "
+        "(except for 3D points) is rank deficient. Number of columns: %d, "
+        "rank: %d.",
+        S_dense.rows(),
+        luOfS_dense.rank());
+    return false;
+  }
+  cov_variables_ = S_dense.inverse();
+  cov_poses_ = cov_variables_.block(0, 0, num_params_poses_, num_params_poses_);
+  return true;
+}
+
+bool BundleAdjustmentCovarianceEstimator::Compute() {
+  if (!HasValidSchurComplement()) {
+    ComputeSchurComplement();
+  }
+
+  // Schur elimination on other variables to get pose covariance
+  Eigen::MatrixXd S_dense = S_matrix_;
+  LOG(INFO) << StringPrintf("Schur elimination on other variables (n = %d)",
+                            num_params_other_variables_);
+  Eigen::MatrixXd S_aa =
+      S_dense.block(0, 0, num_params_poses_, num_params_poses_);
+  Eigen::MatrixXd S_ab = S_dense.block(
+      0, num_params_poses_, num_params_poses_, num_params_other_variables_);
+  Eigen::MatrixXd S_ba = S_ab.transpose();
+  Eigen::MatrixXd S_bb = S_dense.block(num_params_poses_,
+                                       num_params_poses_,
+                                       num_params_other_variables_,
+                                       num_params_other_variables_);
+  Eigen::LDLT<Eigen::MatrixXd> ldltOfS_bb(S_bb);
+  Eigen::MatrixXd S_poses = S_aa - S_ab * ldltOfS_bb.solve(S_ba);
+
+  // Step 5: Compute covariance
+  LOG(INFO) << StringPrintf(
+      "Inverting Schur complement for pose parameters (n = %d)",
+      num_params_poses_);
+  Eigen::FullPivLU<Eigen::MatrixXd> luOfS_poses(S_poses);
+  if (luOfS_poses.rank() < S_poses.rows()) {
+    LOG(INFO) << StringPrintf(
+        "Unable to compute covariance. The Schur complement on pose parameters "
```

Comment:
How well does this approach work when we also estimate intrinsics and some of the (shared) intrinsics variables are ill-conditioned?

Reply:
Yes. But generally the issue happens only for the points. The constraints on the intrinsics are generally pretty dense since it comes from all the residuals. Nonetheless, how about we adding damping factors to all the matrices (rather than only the point blocks) to avoid any potential failures? I can do this if agreed.

Reply:
This can potentially be problematic for parameters with small scale (e.g., the distortion coefficient k1). Maybe we need a more aggressive damping factor there. But I would vote for keeping it like this and return False if the intrinsincs are rank deficient

Reply:
I added the damping factor to the blocks of other variables when we only need pose covariances here https://github.com/colmap/colmap/pull/2610/commits/b6691bf7e3c88150a055d957d5a38dbf23a42e83. This does not alter the results much numerically in my Python test.